opentelemetry-sdk/src/logs/growable_array.rs

```diff
@@ -0,0 +1,323 @@
+        } else if let Some(ref additional) = self.additional {
+            additional.get(index - MAX_STACK_CAPACITY)
+        } else {
+            None
+        }
+    }
+
+    /// Returns the number of elements in the `GrowableArray`.
+    pub(crate) fn len(&self) -> usize {
+        self.count + self.additional.as_ref().map_or(0, Vec::len)
+    }
+
+    /// Returns an iterator over the elements in the `GrowableArray`.
+    ///
+    /// The iterator yields elements from the internal array (`initial`) first, followed by elements
+    /// from the vector (`additional`) if present. This allows for efficient iteration over both
+    /// stack-allocated and heap-allocated portions.
+    ///
+    #[inline]
+    pub(crate) fn iter(&self) -> GrowableArrayIter<'_, T, MAX_STACK_CAPACITY> {
+        if self.additional.is_none() || self.additional.as_ref().unwrap().is_empty() {
+            GrowableArrayIter::StackOnly {
+                iter: self.initial.iter().take(self.count),
+            }
+        } else {
+            GrowableArrayIter::Mixed {
+                stack_iter: self.initial.iter().take(self.count),
+                heap_iter: self.additional.as_ref().unwrap().iter(),
+            }
+        }
+    }
+}
+
+// Implement `IntoIterator` for `GrowableArray`
+impl<T: Default + Clone + PartialEq, const INITIAL_CAPACITY: usize> IntoIterator
+    for GrowableArray<T, INITIAL_CAPACITY>
+{
+    type Item = T;
+    type IntoIter = GrowableArrayIntoIter<T, INITIAL_CAPACITY>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        if self.additional.is_none() || self.additional.as_ref().unwrap().is_empty() {
+            GrowableArrayIntoIter::StackOnly {
+                iter: self.initial.into_iter().take(self.count),
+            }
+        } else {
+            GrowableArrayIntoIter::Mixed {
+                stack_iter: self.initial.into_iter().take(self.count),
+                heap_iter: self.additional.unwrap().into_iter(),
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+/// Iterator for consuming a `GrowableArray`.
+///
+/// The iterator consumes the `GrowableArray`, yielding each element by value.
+/// It first iterates over the elements in the internal array (`initial`), and then,
+/// if present, over the elements in the vector (`additional`).
+///
+pub(crate) enum GrowableArrayIntoIter<T: Default + Clone + PartialEq, const INITIAL_CAPACITY: usize>
+{
+    /// stackonly
+    StackOnly {
+        /// iter
+        iter: std::iter::Take<IntoIter<T, INITIAL_CAPACITY>>,
+    },
+    /// hybrid
+    Mixed {
+        /// stack_iter
+        stack_iter: std::iter::Take<IntoIter<T, INITIAL_CAPACITY>>,
+        /// heap_iter
+        heap_iter: std::vec::IntoIter<T>,
+    },
+}
+
+impl<T: Default + Clone + PartialEq, const INITIAL_CAPACITY: usize> Iterator
+    for GrowableArrayIntoIter<T, INITIAL_CAPACITY>
+{
+    type Item = T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self {
```

Comment:
It's possible to avoid creating an `enum` and matching on its variants on every `next()` call. We could have the required iterator transform ready in `into_iterator` so we don't have to keep checking the presence of elements in vector each time. I tried something like this:
```rust
// Implement `iter` method for `GrowableArray`
#[inline]
pub(crate) fn iter(&self) -> impl Iterator<Item = &T> {
    if self.additional.is_none() || self.additional.as_ref().unwrap().is_empty() {
        self.initial.iter().take(self.count).chain([].iter()) // Chaining with an empty array
         // so that both `if` and `else` branch return the same type
    } else {
        self.initial.iter().take(self.count).chain(self.additional.as_ref().unwrap().iter())
    }
}
```


```rust
// Implement `IntoIterator` for `GrowableArray`
impl<T: Default + Clone + PartialEq, const INITIAL_CAPACITY: usize> IntoIterator
    for GrowableArray<T, INITIAL_CAPACITY>
{
    type Item = T;
    type IntoIter = GrowableArrayIntoIter<T, INITIAL_CAPACITY>;

    fn into_iter(self) -> Self::IntoIter {
        GrowableArrayIntoIter::<T, INITIAL_CAPACITY>::new(self)
    }
}

#[derive(Debug)]
pub(crate) struct GrowableArrayIntoIter<T: Default + Clone + PartialEq, const INITIAL_CAPACITY: usize>
{
    iter: std::iter::Chain<std::iter::Take<std::array::IntoIter<T, INITIAL_CAPACITY>>, std::vec::IntoIter<T>>,
}

impl<T: Default + Clone + PartialEq, const INITIAL_CAPACITY: usize> GrowableArrayIntoIter<T, INITIAL_CAPACITY> {
    fn new(source: GrowableArray<T, INITIAL_CAPACITY>) -> Self{
        Self {
            iter: Self::get_iterator(source)
        }
    }

    fn get_iterator(source: GrowableArray<T, INITIAL_CAPACITY>) -> std::iter::Chain<std::iter::Take<std::array::IntoIter<T, INITIAL_CAPACITY>>, std::vec::IntoIter<T>> {
        if source.additional.is_none() || source.additional.as_ref().unwrap().is_empty() {
            source.initial.into_iter().take(source.count).chain(Vec::<T>::new().into_iter())
        } else {
            source.initial.into_iter().take(source.count).chain(source.additional.unwrap().into_iter())
        }
    }
}

impl<T: Default + Clone + PartialEq, const INITIAL_CAPACITY: usize> Iterator for GrowableArrayIntoIter<T, INITIAL_CAPACITY> {
    type Item = T;

    fn next(&mut self) -> Option<Self::Item> {
        self.iter.next()
    }
}
```

Reply:
Very nice. Thanks for trying this out. This improves the iteration cost. Have incorporated these changes.